### DISCUSSION TITLE
How to do a SQL query with current datetime in UTC?

### DISCUSSION THREAD

**Discussion by tropcicstefan** [OrchardCMS/OrchardCore#8126] [Q&A] [answered]
Let's say there is an EventIndex and I need a list of forthcoming events. IndexProvider stores datetime in UTC. How would I make this SQL query inside orchard?

`SELECT * FROM EventIndex WHERE CURRENT_TIMESTAMP < EventUtc`  - CURRENT_TIMESTAMP doesn't work in orchard

`SELECT * FROM EventIndex WHERE '{{ "now" | date: "%Y-%m-%d %H:%M" }}' < EventUtc` - returns system server datetime

`SELECT * FROM EventIndex WHERE '{{ "DateTime" | shape_new: utc: now, format: "yyyy-MM-dd HH:mm" | shape_stringify }}' < EventUtc` - returns orchard datetime

So the question is what sql function or liquid should I write to get current datetime so it renders into utc, because with given examples rendered result are not comparable with stored utc datetimes in database.

**Comment by Skrypt** [comment]
`{{ "now" | date: "%Y-%m-%d %H:%M" }}` should be returning the Utc datetime.
I'm surprised that it returns the system datetime.

The "DateTime" shape needs that we pass Utc Datetime.
Maybe we made a mistake, I will take a look.

**Reply by Skrypt** [reply]
Ok we need to add a Liquid filter that will return _localClock.ConvertToUtcAsync(value) assuming that the value passed is a Datetime with offset.

Something like :   

`{{ "now" | utc }}`

We need to use the IClock.UtcNow to retrieve it. But if we want to convert a DatetimeOffset to a Utc Datetime then we need to use _localClock.ConvertToUtcAsync(value).

**Comment by Skrypt** [comment, marked as answer]
Ok the correct answer with Liquid; it is using : 

```Liquid
{{ "now" | time_zone: "UTC" }}
```

Here is some test values :  

```Liquid
<p>
	Site Settings timezone set to GMT-7<br/>
	System OS timezone set to GMT-5
</p>
<table>
  <tr class="row">
    <td class="col">
      "now" | local :
    </td>
    <td class="col">
      {{ "now" | local }}
    </td>
    <td>This is the local time based on current tenant site settings</td>
  </tr>
  <tr class="row">
    <td class="col">
      "now" | date: "%Y-%m-%d %H:%M" :
    </td>
    <td class="col">
      {{ "now" | date: "%Y-%m-%d %H:%M" }}
    </td>
    <td>This is the local time based on current OS timezone</td>
  </tr>
  <tr class="row">
    <td class="col">
      "now" | date: "%l:%M%P" :
    </td>
    <td class="col">
      {{ "now" | date: "%l:%M%P" }}
    </td>
    <td>This is the local time based on current OS timezone</td>
  </tr>
  <tr class="row">
    <td class="col">
      "now" | time_zone: "UTC" :
    </td>
    <td class="col">
      {{ "now" | time_zone: "UTC" }}
    </td>
  </tr>
  <tr class="row">
    <td class="col">
      "now" | time_zone: "Europe/London" :
    </td>
    <td class="col">
      {{ "now" | time_zone: "Europe/London" }}
    </td>
  </tr>
  <tr class="row">
    <td class="col">
      "now" | local | time_zone: "Europe/London" :
    </td>
    <td class="col">
      {{ "now" | local | time_zone: "Europe/London" }}
    </td>
  </tr>
  <tr class="row">
    <td class="col">
      "2020-05-18T02:13:09+00:00" | time_zone: "Europe/London" :
    </td>
    <td class="col">
      {{ "2020-05-18T02:13:09+00:00" | time_zone: "Europe/London" }}
    </td>
  </tr>
</table>
```

Here is the result : 

![image](https://user-images.githubusercontent.com/3228637/107797282-c5d32280-6d28-11eb-9022-800b6f557701.png)

So the SQL Query template would look like this : 

```Liquid
SELECT * FROM EventIndex WHERE '{{ "now" | time_zone: "UTC" }}' < EventUtc
```

Alternatively when this PR https://github.com/OrchardCMS/OrchardCore/pull/8557 will be merged we will be able to use the `Now()` function directly with a SQL script.

```SQL
SELECT Now();
-- or
SELECT * FROM EventIndex WHERE Now() < EventUtc
```
